omod/src/main/java/org/openmrs/module/spa/SpaResourceLoader.java

```diff
@@ -0,0 +1,40 @@
+package org.openmrs.module.spa;
+
+import org.springframework.core.io.ContextResource;
+import org.springframework.core.io.DefaultResourceLoader;
+import org.springframework.core.io.FileSystemResource;
+import org.springframework.core.io.Resource;
+import org.springframework.lang.NonNull;
+
+import java.nio.file.Paths;
+
+/**
+ * This is a resource loader for files served by the {@link SpaController}. It is modelled on Spring's
+ * {@link FileSystemContextResource}, but uses absolute paths. Safety checks for where the SPA is served from are handled
+ * by the {@link SpaDirectoryResolver} directly.
+ */
+public class SpaResourceLoader extends DefaultResourceLoader {
```

Comment:
As long as this is being overhauled, I wonder if you considered moving these to runtime properties instead of global properties.  These really don't need to be changed at runtime, and are more environment-specific to how the system is deployed than specific to how the application is configured, so might be more appropriate as runtime properties.  Runtime properties also have the benefit of being very fast to access.

You could also choose to have the best of both worlds, by using the `ConfigUtil` class that I just remembered that I added a few years back to core.  This provides access to runtime, system, and global properties and has a built-in cache that utilizes the global property listener.  It also has a useful `getProperty(String)` method that will check system, then runtime, then global properties (in that order) for a value, so you could support providing values in multiple places, or migrate from one place to another seamlessly.

Reply:
The `ConfigUtil` sounds like a better solution. Although this is a breaking change, I didn't want to break things too much, so I was trying to keep the feature of specifying the directory the same.

Reply:
Alright, so, the only released versions of core that include `ConfigUtil` are 2.3.6 and 2.6.x. (it was backported to 2.4.x and 2.5.x, but we haven't done a release of those platform since). I really kind of want this to be runnable with at least 2.5, so it looks like I can't use that option yet.

Reply:
In either case, I've added support for both runtime and system properties. I don't really expect this value to ever be changed live; my motivation was just to ensure that anyone who was already using this and had customized their frontend folder wouldn't need to make any changes.